### PR TITLE
KAFKA-6049: Add session window support for cogroup

### PR DESCRIPTION
**Pull request by wcarlson5**
adding SessionWindowedCogroupedKStream options to the cogroup operator

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java [hunk 1]
```diff
@@ -264,7 +264,14 @@
     KTable<K, VOut> aggregate(final Initializer<VOut> initializer,
                               final Named named,
                               final Materialized<K, VOut, KeyValueStore<Bytes, byte[]>> materialized);
-
+    /**
+     * Create a new {@link SessionWindowedCogroupedKStream} instance that can be used to perform session
+     * windowed aggregations.
+     *
+     * @param sessionWindows the specification of the aggregation {@link SessionWindows}
+     * @return an instance of {@link TimeWindowedKStream}
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java]: return type is `SessionWindowedCogroupedKStream`

File: streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedCogroupedKStream.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.state.SessionStore;
+/**
+ * {@code SessionWindowedCogroupKStream} is an abstraction of a <i>windowed</i> record stream of {@link org.apache.kafka.streams.KeyValue} pairs.
+ * It is an intermediate representation of a {@link CogroupedKStream} in order to apply a windowed aggregation operation on the original
+ * {@link KGroupedStream} records.
+ * <p>
+ * The specified {@link SessionWindows} defines how the windows are created.
+ * The result is written into a local windowed {@link org.apache.kafka.streams.state.KeyValueStore} (which is basically an ever-updating
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedCogroupedKStream.java]: `local {@link SessionStore}`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    private final SessionWindows sessionWindows;
+    private final CogroupedStreamAggregateBuilder<K, V> aggregateBuilder;
+    private final Map<KGroupedStreamImpl<K, ?>, Aggregator<? super K, ? super Object, V>> groupPatterns;
+
+    SessionWindowedCogroupedKStreamImpl(final SessionWindows sessionWindows,
+                                        final InternalStreamsBuilder builder,
+                                        final Set<String> sourceNodes,
+                                        final String name,
+                                        final Serde<K> keySerde,
+                                        final Serde<V> valSerde,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java]: remove keySerde and valSerde as both are not used.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+import org.apache.kafka.streams.kstream.internals.graph.StreamsGraphNode;
+import org.apache.kafka.streams.state.SessionBytesStoreSupplier;
+import org.apache.kafka.streams.state.SessionStore;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.apache.kafka.streams.state.Stores;
+
+public class SessionWindowedCogroupedKStreamImpl<K, V> extends
+    AbstractStream<K, V> implements SessionWindowedCogroupedKStream<K, V> {
+
+    private static final String AGGREGATE_NAME = "COGROUPKSTREAM-AGGREGATE-";
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java]: See duplicates `CogroupedKStreamImpl#AGGREGATE_NAME` (actually same issue in already merged `TimeWindowedCogroupedKStreamImpl` that also duplicate is -- can you fix both?)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        //keySerde and valueSerde are null because there are many different groupStreams that they could be from
+        this.sessionWindows = sessionWindows;
+        this.aggregateBuilder = aggregateBuilder;
+        this.groupPatterns = groupPatterns;
+    }
+
+    @Override
+    public KTable<Windowed<K>, V> aggregate(final Initializer<V> initializer,
+                                            final Merger<? super K, V> sessionMerger) {
+        return aggregate(initializer, sessionMerger, Materialized.with(keySerde, null));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java]: the `keySerde` is set to `null` in the constructor, hence, I am wondering if we should pass in `null`, here, too? (similar question for `TimeWindowedCogroupedKStreamImpl`)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+                                            final Merger<? super K, V> sessionMerger,
+                                            final Materialized<K, V, SessionStore<Bytes, byte[]>> materialized) {
+        return aggregate(initializer, NamedInternal.empty(), sessionMerger, materialized);
+    }
+
+    @Override
+    public KTable<Windowed<K>, V> aggregate(final Initializer<V> initializer,
+                                            final Named named,
+                                            final Merger<? super K, V> sessionMerger) {
+        return aggregate(initializer, named, sessionMerger, Materialized.with(keySerde, null));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java]: as above.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+    @Override
+    public KTable<Windowed<K>, V> aggregate(final Initializer<V> initializer,
+                                            final Named named,
+                                            final Merger<? super K, V> sessionMerger,
+                                            final Materialized<K, V, SessionStore<Bytes, byte[]>> materialized) {
+        Objects.requireNonNull(initializer, "initializer can't be null");
+        Objects.requireNonNull(sessionMerger, "sessionMerger can't be null");
+        Objects.requireNonNull(materialized, "materialized can't be null");
+        Objects.requireNonNull(named, "named can't be null");
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java:88]: @bbejeck -- I was double checking existing code and many (all) operator actually allow `named` to be `null` -- do you remember why? I am wondering if this is a issue with existing code or an issue with the new code?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedCogroupedKStreamImpl.java [hunk 8]
```diff
@@ -93,7 +93,7 @@
         return aggregateBuilder.build(
             groupPatterns,
             initializer,
-            NamedInternal.empty(),
+            new NamedInternal(named),
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedCogroupedKStreamImpl.java:95]: Why this change?

File: streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedCogroupedKStream.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+     * @param initializer    the instance of {@link Initializer}. Cannot be {@code null}.
+     * @param named  an instance of {@link Named} used to Named the processors. Cannot be {@code null}.
+     * @param sessionMerger  the instance of {@link Merger}. Cannot be {@code null}.
+     * @param materialized   an instance of {@link Materialized} used to materialize a state store. Cannot be {@code null}.
+     * @return a windowed {@link KTable} that contains "update" records with unmodified keys, and values that represent
+     * the latest (rolling) aggregate for each key within a window
+     */
+    KTable<Windowed<K>, V> aggregate(final Initializer<V> initializer,
+                                     final Named named,
+                                     final Merger<? super K, V> sessionMerger,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedCogroupedKStream.java:155]: `sessionMerger` should be the second argument

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImplTest.java [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+            assertOutputKeyValueTimestamp(testOutputTopic, "k2", "0+B", 607);
+        }
+    }
+
+    @Test
+    public void sessionWindowAggregate2Test() {
+        final KTable<Windowed<String>, String> customers = groupedStream.cogroup(MockAggregator.TOSTRING_ADDER)
+                .windowedBy(SessionWindows.with(ofMillis(500)))
+                .aggregate(MockInitializer.STRING_INIT, sessionMerger, Materialized.with(Serdes.String(), Serdes.String()));
+        customers.toStream().to(OUTPUT);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImplTest.java:186]: This seems to be the same program as in the test above -- if yes, I think we should merge both tests into one.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedCogroupedKStreamImpl.java (hunk 8 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedCogroupedKStreamImpl.java:95]: Something I missed when updating TimeWindowed which could brake the naming for a few cases

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImplTest.java (hunk 10 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImplTest.java:186]: its a small difference, this one has a merge where the one above does not

File: streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java (hunk 1 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java]: I had it right the first time...

File: streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedCogroupedKStream.java (hunk 2 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedCogroupedKStream.java]: looks like I missed one...

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java (hunk 4 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java]: Yeah, both changed

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java (hunk 5 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java]: its probably less confusing

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java (hunk 7 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java:88]: Since there are options to just not pass it in anyone who would put null should just not include it

Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java:88]: >@bbejeck -- I was double checking existing code and many (all) operator actually allow named to be null -- do you remember why? I am wondering if this is a issue with existing code or an issue with the new code?

Which operators are you referring to? I just took a quick look at `KStreamImpl` and they don't allow null `Named`.  Several of the overloads don't take a `Named` and create a `NamedInternal#empty`

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedCogroupedKStreamImpl.java:88]: There are a couple of examples, but your comment indicates, that they should allow it. Just wanted to confirm. Will go over the code and add the corresponding checks in a follow up PR to fix it.

**Comment by mjsax** [comment]
Java 8 / 2.11 failed:
```
kafka.api.DelegationTokenEndToEndAuthorizationTest.testNoConsumeWithDescribeAclViaAssign
org.apache.kafka.streams.integration.OptimizedKTableIntegrationTest.shouldApplyUpdatesToStandbyStore
```
Java 11 passed.

Retest this please.
